FieldOffloadBlockedTransformation - Add transformation for blocked and asynchronous data offload using FIELD API

wertysas opened the pull request:
This PR adds a new transformation, FieldOffloadBlockedTransformation, that can block driver loops and perform blocked (optionally asynchronous) offload using Field API.

There is a corresponding dwarf-p-cloudsc PR:

wertysas commented:
I have rebased the contents of this PR over the LoopTree PR [#562](https://github.com/ecmwf-ifs/loki/pull/562) (on a separate branch: [``je-field-offload-blocking-loop-tree-rebase``](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase)) and manually triggered the [tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492522430), [regression-tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492530054), and [code-checks](https://github.com/ecmwf-ifs/loki/actions/runs/15492510660) workflows which all pass.

Hence, I will open this PR now as well, and if #562  gets accepted I can point this branch to the rebased version, but I am keeping it at is for now in case we would prefer another implementation of the ``find_driver_loop`` utility.

wertysas commented:
> I have rebased the contents of this PR over the LoopTree PR [#562](https://github.com/ecmwf-ifs/loki/pull/562) (on a separate branch: [`je-field-offload-blocking-loop-tree-rebase`](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase)) and manually triggered the [tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492522430), [regression-tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492530054), and [code-checks](https://github.com/ecmwf-ifs/loki/actions/runs/15492510660) workflows which all pass.
> 
> Hence, I will open this PR now as well, and if #562 gets accepted I can point this branch to the rebased version, but I am keeping it at is for now in case we would prefer another implementation of the `find_driver_loop` utility.


@reuterbal @mlange05 Since #562  has been accepted now I have updated the rebase over the final version of  #562 and checked that the test/code-check/regression-test all pass. Have you started reviewing this yet (I don't want to force push the branch if someone is in the middle of a review) **otherwise I will repoint this branch to  [`je-field-offload-blocking-loop-tree-rebase`](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase).**

reuterbal commented:
> Have you started reviewing this yet (I don't want to force push the branch if someone is in the middle of a review) **otherwise I will repoint this branch to [`je-field-offload-blocking-loop-tree-rebase`](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase).**

All good from my side, feel free to push the updated branch!

wertysas commented:
> > ~I have rebased the contents of this PR over the LoopTree PR [#562](https://github.com/ecmwf-ifs/loki/pull/562) (on a separate branch: [`je-field-offload-blocking-loop-tree-rebase`](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase)) and manually triggered the [tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492522430), [regression-tests](https://github.com/ecmwf-ifs/loki/actions/runs/15492530054), and [code-checks](https://github.com/ecmwf-ifs/loki/actions/runs/15492510660) workflows which all pass.
> > Hence, I will open this PR now as well, and if #562 gets accepted I can point this branch to the rebased version, but I am keeping it at is for now in case we would prefer another implementation of the `find_driver_loop` utility.~
> 
> ~@reuterbal @mlange05 Since #562 has been accepted now I have updated the rebase over the final version of #562 and checked that the test/code-check/regression-test all pass. Have you started reviewing this yet (I don't want to force push the branch if someone is in the middle of a review) **otherwise I will repoint this branch to [`je-field-offload-blocking-loop-tree-rebase`](https://github.com/ecmwf-ifs/loki/tree/je-field-offload-blocking-loop-tree-rebase).**~

@reuterbal @mlange05 I have rebased over #562 now and this is ready for review.

wertysas commented:
@reuterbal @mlange05 This is ready for review!

I have rebased onto main following the merge of #577 and verified that the asynchronous implementation works with cloudsc (see output below).

During the rebase I had to resolve conflicts with changes made to the FIELD API utilities in #564. Specifically, there was a test, ``test_field_api_call_gen``, added to ``test_field_offload.py`` in #564, that was testing new field offload utilities from ``field_api.py``. However, there were already existing tests for the utilities in ``field_api.py`` located in ``transformations/parallel/tests/test_field_views.py`` and therefore I moved (and modified due to overlap with existing tests) the tests added in #564 to ``test_field_views.py``.


### Results from a cloudsc run on ATOS on an A100 node with 720000 points

(The increase in performance compared to the graph in the PR description comes from the recent SLURM and configuration update on ATOS which leads to better performance results across all cloudsc variants on ATOS)

```
PGI_ACC_CUDA_HEAPSIZE=8GB srun --nodes=1 --mem=200G --gres=gpu:1 -q dg bin/dwarf-cloudsc-loki-scc-field-async-dp 1 720000 128
srun: job 31769631 queued and waiting for resources
srun: job 31769631 has been allocated resources
     NUMPROC=1, NUMOMP=1, NGPTOTG=720000, NPROMA=128, NGPBLKS=5625
 Reference MFLOP count for 100 columns :  12.48232900
     NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s     col/s
          1    720000    720000         1       128    0 :       2557     35141    281529 @ rank#0:core#0
          1    720000    720000      5625       128   -1 :       2557     35141    281527 : TOTAL @ rank#0
    1 x   1    720000    720000      5625       128   -1 :       2557     35141    281527 : TOTAL
 DEV_ALLOC_SIZE = 68719476736
             Variable Dim             MinValue             MaxValue            AbsMaxErr         AvgAbsErr/GP          MaxRelErr-%
                PLUDE 2D1  0.0000000000000E+00  0.1026720108982E-03  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
             PCOVPTOT 2D3  0.0000000000000E+00  0.1000000000000E+01  0.3108624468950E-14  0.6721012635325E-15  0.2522121455387E-14
     PRAINFRAC_TOPRFZ 1D1  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
               PFSQLF 2D3 -0.2581756360096E-05  0.4080862802487E-05  0.1355682849525E-18  0.1956214325263E-18  0.1921048884992E-11 !!!!
               PFSQIF 2D3 -0.2810873821405E-05  0.3887836598167E-05  0.1687713147404E-18  0.1866589088139E-18  0.8329865159329E-12 !!!!
              PFCQLNG 2D1 -0.3555564290808E-07  0.1413389467871E-08  0.3308722450212E-23  0.4230994440106E-23  0.0000000000000E+00
              PFCQNNG 2D1 -0.5285014021662E-06  0.8788562329576E-08  0.5293955920339E-22  0.2312968244934E-22  0.0000000000000E+00
               PFSQRF 2D3 -0.2581756360096E-05  0.4080862802487E-05  0.1355682849525E-18  0.1883884680564E-18  0.1948114895849E-11 !!!!
               PFSQSF 2D3 -0.2768876986308E-05  0.3860349895496E-05  0.1759710947921E-18  0.1859606909121E-18  0.8346660548130E-12 !!!!
              PFCQRNG 2D1 -0.3555564290808E-07  0.1413389467871E-08  0.3308722450212E-23  0.4029417633410E-23  0.0000000000000E+00
              PFCQSNG 2D1 -0.5285014021662E-06  0.8788562329576E-08  0.5293955920339E-22  0.2107105285572E-22  0.0000000000000E+00
             PFSQLTUR 2D1 -0.8115949343062E-06  0.2653456385227E-06  0.1058791184068E-21  0.3489066860269E-22  0.0000000000000E+00
             PFSQITUR 2D1 -0.2651690301090E-05  0.3864567769232E-06  0.2117582368136E-21  0.1205586209974E-21  0.0000000000000E+00
               PFPLSL 2D1  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
               PFPLSN 2D3  0.0000000000000E+00  0.2969958004108E-04  0.2846030702774E-18  0.3417666893820E-18  0.1973617838346E-12
               PFHPSL 2D1 -0.0000000000000E+00 -0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
               PFHPSN 2D3 -0.8418345962643E+02 -0.0000000000000E+00  0.8064660050877E-12  0.9693275054964E-12  0.1974819492896E-12
       TENDENCY_LOC%A 2D3 -0.2777777777778E-03  0.2631309809623E-03  0.6965998958219E-17  0.1001500487235E-17  0.2193923791312E-12
       TENDENCY_LOC%Q 2D3 -0.1997382404851E-07  0.1495580073796E-07  0.1747005453712E-20  0.5554307276100E-21  0.7090881617711E-12 !!!!
       TENDENCY_LOC%T 2D3 -0.4132200394167E-04  0.5435278446361E-04  0.4919567357653E-17  0.1916666398482E-17  0.8624917371582E-12 !!!!
     TENDENCY_LOC%CLD 3D3 -0.1222481036148E-07  0.9816546350973E-08  0.5835112986708E-21  0.3403818904211E-21  0.8529854197599E-12 !!!!

```

reuterbal reviewed (changes requested):
Many thanks, this is a really powerful addition. The implementation is well-structured and broken down into logical steps with appropriate utilities. But it also encapsulates a lot of bespoke IFS-Field-API knowledge, which isn't possible to follow without the relevant context. So, maybe we can make this a little more accessible with a few in-line comments that explain the rationale behind some of the steps?

Otherwise I've left a few remarks how certain implementation details might be possible to improve

wertysas commented on loki/transformations/loop_blocking.py:
```diff
@@ -131,11 +141,29 @@ def split_loop(routine: Subroutine, loop: ir.Loop, block_size: int):
-    change_map = {loop: block_loop_inits + (outer_loop,)}
-    Transformer(change_map, inplace=True).visit(routine.body)
+    #  Create outer_loop with blocking vars and inner_loop in its body
+    if data_region is None:
+        outer_loop = loop.clone(variable=splitting_vars.block_idx, body=blocking_body + (inner_loop,),
+                                bounds=sym.LoopRange((sym.IntLiteral(1), splitting_vars.num_blocks)))
+        change_map = {loop: block_loop_inits + (outer_loop,)}
+        Transformer(change_map, inplace=True).visit(routine.body)
+    # If data_region is provided we block the entire data_region and place the
+    # outer loop outside the data region
+    else:
+        outer_loop = loop.clone(variable=splitting_vars.block_idx, body=blocking_body + (data_region,),
+                                bounds=sym.LoopRange((sym.IntLiteral(1), splitting_vars.num_blocks)))
+
+        # Transformer to place block loop outside data region
+        class BlockDataRegionTransformer(Transformer):
+            def visit_PragmaRegion(self, pragma_region):
+                if pragma_region is not data_region:
+                    return pragma_region
+                change_map = {loop: (inner_loop,)}
+                Transformer(change_map, inplace=True).visit(data_region)
+                return block_loop_inits + (outer_loop,)
+
+        BlockDataRegionTransformer(inplace=True).visit(routine.body)
```
The reason for this is that the pragmas are not part of the region and in this case we need the block loop to be placed outside the data region. I did something similar to what you suggested at first but this resulted in the data pragmas being placed outside the loop resulting in an incorrect data region for subsequent transformations.

I also think it feels a bit chunky to introduce a local transformer here for this, but it was the easiest solution that came to my mind when implementing this.

reuterbal replied:
OK, I had another look at this and the problem should be solvable. There are two things to work around here:

1. You are using the data region, outer loop and inner loop nodes across several transformation steps, thus the rebuild needs to be in-place always to make sure these references stay up-to-date. You have been doing this diligently.
2. Because these nodes are always the same object, the in-place transformer pass over routine.body replaces data_region with itself wrapped into something - and incurs infinite recursion.

The easiest way that I found to solve both was to accept a partial rebuild for the data region, and to return that in split_loops to ensure it is up-to-date on the caller side. With this, tests seem to be passing but it is of course likely that there is still something foul. I'll file a PR against your branch with a commit that has my changes.

wertysas commented on loki/transformations/field_api.py:
```diff
@@ -159,52 +241,118 @@ def field_get_device_data(field_ptr, dev_ptr, transfer_type: FieldAPITransferTyp
+    if transfer_type != FieldAPITransferType.FORCE and (queue is not None or blk_bounds is not None):
+        raise ValueError("Only force copy methods can have non-None type queue or blk_bounds")
+
     if transfer_type == FieldAPITransferType.READ_ONLY:
```
Yes, I went even further and placed the string representation as the value property on the ENUM types and added a property ``suffix`` returning that returns this value. This simplifies the ``_field_sync`` and ``_field_get`` functions even further.

wertysas commented:
@reuterbal Thank you for the comments,  I have added the missing docstrings and in addition to your suggested changes I have also:
* Renamed ``replace_kernel_args`` to ``replace_offload_args`` in ``field_offload.py``
* Added the string representations of the ENUM types in ``field_api.py`` to the types as values, and added a ``suffix`` property to make this value easily accessible in the methods using them.

reuterbal reviewed (changes requested):
Many thanks for the changes! This is almost ready, but I'm still not convinced by the Transformer pattern - so I had a go myself and filed the suggested changes as #588 against your PR branch.

mlange05 reviewed (approved):
A very powerful addition and very neatly implemented! Many thanks, GTG from me! 👏 🙏 :shipit:
